### PR TITLE
Add nodes for spatial and temporal merge

### PR DESCRIPTION
**Pull request by msz-rai**

### CODE REVIEW

File: src/graph/TemporalMergePointsNode.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+void TemporalMergePointsNode::schedule(cudaStream_t stream)
+{
+	for (const auto& [field, data] : mergedData) {
+		size_t pointCount = input->getPointCount();
+		const auto toMergeData = input->getFieldData(field, stream);
+		data->insertData(toMergeData->getReadPtr(MemLoc::Device), pointCount, width);
+	}
+	width += input->getWidth();
```
Comment (prybicki) [inline src/graph/TemporalMergePointsNode.cpp:61]: Data must be merged on the Host side. We are already running short of GPU memory with simulators.

**Review by prybicki — CHANGES_REQUESTED** [review]

**Review by PiotrMrozikRobotec — DISMISSED** [review]
Generally speaking, I would consider the idea of implementing different merging methods this way. We can talk about DRY, and how to avoid dozens of merging methods in the future. Maybe some general merge method with parameters and specializations? To think about.

File: src/VArray.hpp [hunk 2]
```diff
@@ -79,6 +79,7 @@ struct VArray : std::enable_shared_from_this<VArray>
 	const void* getReadPtr(MemLoc location) const;
 	void setData(const void* src, std::size_t elements);
 	void getData(void* dstRaw, std::size_t count) const;
+	void insertData(const void *src, std::size_t elements, std::size_t offset);
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/VArray.hpp:82]: DRY, [ditto](https://pl.wiktionary.org/wiki/ditto)

File: src/VArrayProxy.hpp [hunk 3]
```diff
@@ -37,6 +37,7 @@ struct VArrayProxy
 
 	void setData(const T* srcRaw, std::size_t count) { src->setData(srcRaw, count); }
 	void getData(T* dstRaw, std::size_t count) const { src->getData(dstRaw, count); }
+	void insertData(const T* srcRaw, std::size_t count, std::size_t offset) { src->insertData(srcRaw, count, offset); }
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/VArrayProxy.hpp:40]: 1) Comprehensive and non-trivial comments are gold
2) Trivial comments have a negative value of occupying precious screen space
3) This particular method, along with its parameter, is IMVHO obvious to any C++ programmer (it is analogous to `std::vector<T>::insert`).
4) Also, this class has the status `to be removed` at the nearest opportunity. We will include this topic in our in-person meeting.

File: src/api/apiCore.cpp [hunk 4]
```diff
@@ -655,6 +655,54 @@ void TapePlayer::tape_node_points_write_pcd_file(const YAML::Node& yamlNode)
...
+	return status;
+}
+
+void TapePlayer::tape_node_points_spatial_merge(const YAML::Node& yamlNode)
+{
+	size_t nodeId = yamlNode[0].as<TapeAPIObjectID>();
+	rgl_node_t node = tapeNodes.contains(nodeId) ? tapeNodes[nodeId] : nullptr;
+	rgl_node_points_spatial_merge(&node,
+		reinterpret_cast<const rgl_field_t*>(fileMmap + yamlNode[1].as<size_t>()),
+		yamlNode[2].as<int>());
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/api/apiCore.cpp:678]: 1. Not sure about this. Changes you have suggested relate to the second param - `rgl_field_t fields`, not the `Node*`. 
2. Good point :+1:

File: src/graph/DownSamplePointsNode.cpp [hunk 5]
```diff
@@ -26,6 +26,10 @@ void DownSamplePointsNode::validate()
 		CHECK_CUDA(cudaEventCreateWithFlags(&finishedEvent, cudaEventDisableTiming));
 	}
 	input = getValidInput<IPointsNode>();
+	if (!input->hasField(XYZ_F32)) {
+		auto msg = fmt::format("{} requires XYZ to be present", getName());
+		throw InvalidPipeline(msg);
+	}
```
[the remark this answers is not in the document]
Comment (PiotrMrozikRobotec) [reply src/graph/DownSamplePointsNode.cpp:32]: As far as this is our internal lib, I would suggest: leave here a comment with that. Leave it for the next generation!

File: src/graph/DownSamplePointsNode.cpp [hunk 6]
```diff
@@ -26,6 +26,10 @@ void DownSamplePointsNode::validate()
 		CHECK_CUDA(cudaEventCreateWithFlags(&finishedEvent, cudaEventDisableTiming));
 	}
 	input = getValidInput<IPointsNode>();
+	if (!input->hasField(XYZ_F32)) {
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/graph/DownSamplePointsNode.cpp:29]: IMHO, this change is related to the merging nodes. There is no guarantee that the user sets proper fields to the `TemporalMerge`/`SpatialMerge`, and this case should be caught in the graph validation.

File: src/graph/DownSamplePointsNode.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/graph/DownSamplePointsNode.cpp:32]: I can iterate over `getRequiredFieldList()`

File: src/graph/Interfaces.hpp [hunk 7]
```diff
@@ -59,12 +59,12 @@ struct IPointsNode
 
 	// Point cloud description
 	virtual bool isDense() const = 0;
-	virtual bool hasField(rgl_field_t field) const = 0;
 	virtual std::size_t getWidth() const = 0;
 	virtual std::size_t getHeight() const = 0;
 	virtual std::size_t getPointCount() const { return getWidth() * getHeight(); }
 
 	// Data getters
+	virtual bool hasField(rgl_field_t field) const = 0;
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/graph/Interfaces.hpp:67]: My bad. Reverted

[the remark this answers is not in the document]
Comment (msz-rai) [reply src/graph/Interfaces.hpp:67]: ditto

File: src/graph/SpatialMergePointsNode.cpp [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <graph/NodesCore.hpp>
+#include <RGLFields.hpp>
+
+void SpatialMergePointsNode::setParameters(const std::vector<rgl_field_t>& fields)
+{
+	if (std::find(fields.begin(), fields.end(), RGL_FIELD_DYNAMIC_FORMAT) != fields.end()) {
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/graph/SpatialMergePointsNode.cpp:20]: Not much different to me, but OK.

File: src/graph/SpatialMergePointsNode.cpp [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+		// Check input pointcloud is unorganized
+		if (input->getHeight() != 1) {
+			auto msg = "Spatial points merge could be proceed on unorganized pointclouds only";
+			throw InvalidPipeline(msg);
+		}
+
+		// Check input pointcloud has required fields
+		for (const auto& [requiredField, _] : mergedData) {
+			if (!input->hasField(requiredField)) {
+				auto msg = fmt::format("SpatialMergePointsNode input has not required field '{}'",
```
[the remark this answers is not in the document]
Comment (PiotrMrozikRobotec) [reply src/graph/SpatialMergePointsNode.cpp:53]: `input has not` -> `input does not have`

File: src/graph/SpatialMergePointsNode.cpp [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+	}
+}
+
+bool SpatialMergePointsNode::isDense() const
+{
+	auto notDense = [](IPointsNode::Ptr node) { return node->isDense() == false; };
+	if (std::any_of(pointInputs.begin(), pointInputs.end(), notDense)) {
+		return false;
+	}
+	return true;
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/graph/SpatialMergePointsNode.cpp:86]: Good point!

File: src/graph/TemporalMergePointsNode.cpp [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+	// Check input pointcloud is unorganized
+	if (input->getHeight() != 1) {
+		auto msg = "Temporal points merge could be proceed on unorganized pointclouds only";
+		throw InvalidPipeline(msg);
+	}
+
+	// Check input pointcloud has required fields
+	for (const auto& [requiredField, _] : mergedData) {
+		if (!input->hasField(requiredField)) {
+			auto msg = fmt::format("TemporalMergePointsNode input has not required field '{}'",
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/graph/TemporalMergePointsNode.cpp:47]: ditto

File: src/graph/TemporalMergePointsNode.cpp (hunk 1 above)
Comment (msz-rai) [reply src/graph/TemporalMergePointsNode.cpp:61]: Do you think this change is enough?
```suggestion
	for (const auto& [field, data] : mergedData) {
		size_t pointCount = input->getPointCount();
		const auto toMergeData = input->getFieldData(field, stream);
		data->insertData(toMergeData->getReadPtr(MemLoc::Host), pointCount, width);
	}
	width += input->getWidth();
```
Output nodes could use merged data on the GPU, so the data will be transferred anyway.

File: src/graph/TransformPointsNode.cpp [hunk 12]
```diff
@@ -20,7 +20,7 @@ void TransformPointsNode::validate()
 	input = getValidInput<IPointsNode>();
 	if (!input->hasField(XYZ_F32)) {
 		auto msg = fmt::format("{} requires XYZ to be present", getName());
-		throw std::invalid_argument(msg);
+		throw InvalidPipeline(msg);
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/graph/TransformPointsNode.cpp:23]: Leave your code better than you found it. :wink:

**Comment by msz-rai** [comment]
Rebased

File: src/api/apiCore.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (prybicki) [reply src/api/apiCore.cpp]: Yes, 1. is completely wrong, it was a brain fart.

File: src/api/apiCore.cpp [hunk 13]
```diff
@@ -655,6 +655,54 @@ void TapePlayer::tape_node_points_write_pcd_file(const YAML::Node& yamlNode)
...
+}
+
+void TapePlayer::tape_node_points_temporal_merge(const YAML::Node& yamlNode)
+{
+	size_t nodeId = yamlNode[0].as<TapeAPIObjectID>();
+	rgl_node_t node = tapeNodes.contains(nodeId) ? tapeNodes[nodeId] : nullptr;
+	rgl_node_points_temporal_merge(&node,
+		reinterpret_cast<const rgl_field_t*>(fileMmap + yamlNode[1].as<size_t>()),
+		yamlNode[2].as<int>());
+	tapeNodes.insert(std::make_pair(nodeId, node));
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/api/apiCore.cpp]: Resolve if the `int32_t `has been introduced.

**Review by prybicki — APPROVED** [review]

File: src/graph/TemporalMergePointsNode.cpp (hunk 1 above)
Comment (prybicki) [reply src/graph/TemporalMergePointsNode.cpp:62]: You have just changed the source of the data which will needlessly copy input to the host. It is the destination VArray which must reside on the Host. Currently, I'm not able to provide a solution for this (semi-vacation), but I'm okay with merging as-is, provided that you add me a notion task to fix this.

Comment (msz-rai) [reply src/graph/TemporalMergePointsNode.cpp:62]: OK :bowing_man:

**Pull request merged**